#VRML_SIM R2023a utf8
# license: Copyright Cyberbotics Ltd. Licensed for use only with Webots.
# license url: https://cyberbotics.com/webots_assets_license
# documentation url: https://www.cyberbotics.com/doc/automobile/car
# Simple example shwoing how to inherit from AckermannVehicle to build any other vehicle.
# This model was sponsored by the CTI project RO2IVSim (http://transport.epfl.ch/simulator-for-mobile-robots-and-intelligent-vehicles).
# template language: javascript

EXTERNPROTO "AckermannVehicle.proto"

PROTO BuffCar [
  field SFVec3f    translation                    0 0 0.4
  field SFRotation rotation                       0 0 1 0
  field SFString   name                           "vehicle"
  field SFString   controller                     "<generic>"
  field MFString   controllerArgs                 []
  field SFString   window                         "<generic>"
  field SFBool     supervisor                     FALSE
  field SFBool     synchronization                TRUE
  field SFFloat    trackFront                     0.254
  field SFFloat    trackRear                      0.254
  field SFFloat    wheelbase                      0.2921
  field SFFloat    suspensionFrontSpringConstant  100000
  field SFFloat    suspensionFrontDampingConstant 4000
  field SFFloat    suspensionRearSpringConstant   100000
  field SFFloat    suspensionRearDampingConstant  4000
  field SFFloat    minSteeringAngle               -1
  field SFFloat    maxSteeringAngle               1
  field MFNode     bodyExpansionSlot              []
  field MFNode     sensorExpansionSlot            []
  field SFNode     boundingObject                 NULL
  field SFNode     physics                        Physics { density 1000 }
]
{
  AckermannVehicle {
    translation IS translation
    rotation IS rotation
    model "simple vehicle"
    axisDevicesRearLeft [
      RotationalMotor {
        name "rwd_motor_left"
        maxTorque 5000
        maxVelocity 20
      }
    ]
    axisDevicesRearRight [
      RotationalMotor {
        name "rwd_motor_right"
        maxTorque 5000
        maxVelocity 20
      }
    ]
    extensionSlot [
      Group {
        children IS sensorExpansionSlot
      }
      DEF CAR_BODY Transform {
        children IS bodyExpansionSlot
      }
      Transform {
        translation %<= fields.wheelbase.value >% 0 0
        rotation 0.577351 0.577351 0.57735 -2.094393
          children [
            Shape {
              appearance PBRAppearance {
                baseColor 1.0 0.0 0.0
                metalness 0
                roughness 0.6
              }
              geometry Cylinder {
                height %<= fields.trackFront.value >%
                radius 0.004
              }
            }
          ]
        }
        Transform {
          rotation 0.577351 0.577351 0.57735 -2.094393
          children [
            Shape {
              appearance PBRAppearance {
                baseColor 0.0 0.0 1.0
                metalness 0
                roughness 0.6
              }
              geometry Cylinder {
                height %<= fields.trackRear.value >%
                radius 0.004
              }
            }
          ]
        }
        Transform {
          translation %<= fields.wheelbase.value/2 >% 0 0
          rotation 0 1 0 -1.570795
          children [
            Shape {
              appearance PBRAppearance {
                baseColor 0.0 1.0 0.0
                metalness 0
                roughness 0.6
              }
              geometry Cylinder {
                height %<= fields.wheelbase.value >%
                radius 0.006
              }
            }
          ]
        }
        Transform {
          translation %<= fields.wheelbase.value >% 0 0
          rotation -0.57735 -0.57735 0.57735 2.094395
          children [
            Shape {
              appearance PBRAppearance {
                baseColor 0.8 0.8 0.8
                metalness 0
                roughness 0.6
              }
              geometry Box {
                size 0.02 0.02 0.02
              }
            }
          ]
        }
        Shape {
          appearance PBRAppearance {
            baseColor 1 1 0.0
            metalness 0
            roughness 0.6
          }
          geometry Box {
            size 0.02 0.02 0.02
          }
        }
      ] 
    name IS name
    controller IS controller
    controllerArgs IS controllerArgs
    window IS window
    supervisor IS supervisor
    synchronization IS synchronization
    trackFront IS trackFront
    trackRear IS trackRear
    wheelbase IS wheelbase
    suspensionFrontSpringConstant  IS suspensionFrontSpringConstant
    suspensionFrontDampingConstant IS suspensionFrontDampingConstant
    suspensionRearSpringConstant IS suspensionRearSpringConstant
    suspensionRearDampingConstant IS suspensionRearDampingConstant
    minSteeringAngle IS minSteeringAngle
    maxSteeringAngle IS maxSteeringAngle
    boundingObject IS boundingObject
    physics IS physics
  }
}
